File: Coding_1124 Backtracking/최원준/이강훈/2번.java
/*
https://leetcode.com/problems/maximum-length-of-a-concatenated-string-with-unique-characters/description/

1. 아이디어
백트래킹 구현

2. 자료구조
배열

3. 시간 복잡도
O(2^N)
*/

import java.util.List;

public class Solution {
    int max = 0;

    public int maxLength(List<String> arr) {
        backtrack(arr, "", 0);

        return max;
    }

    private void backtrack(List<String> arr, String current, int start) {
        if (max < current.length()) max = current.length();

        for (int i = start; i < arr.size(); i++) {
            if (!isValid(current, arr.get(i))) continue;

            backtrack(arr, current + arr.get(i), i + 1);
        }
    }

    private boolean isValid(String orgString, String newString) {
        int[] array = new int[26];

        for (int i = 0; i < newString.length(); i++) {
            if (++array[newString.charAt(i) - 'a'] == 2) return false;
            if (orgString.contains(String.valueOf(newString.charAt(i)))) return false;
        }

        return true;
    }
}

File: Coding_231124 Backtracking/최원준/최원준/2번.py
# https://leetcode.com/problems/maximum-length-of-a-concatenated-string-with-unique-characters/

'''
1. 아이디어 :
    overlap 함수를 만들어서, 중복되는 문자가 있는지 확인한다.
    없으면, charSet에 추가하고, backtrack(i+1)을 호출한다.
    있으면, backtrack(i+1)을 호출한다.
2. 시간복잡도 :
    O(n^2)
3. 자료구조 :
    해시맵
'''

class Solution:
    def maxLength(self, arr: List[str]) -> int:
        charSet= set()

        def overlap(charSet, s):
            c = Counter(charSet) + Counter(s)
            return max(c.values()) > 1

        def backtrack(i):
            if i == len(arr):
                return len(charSet)

            res = 0
            if not overlap(charSet, arr[i]):
                for c in arr[i]:
                    charSet.add(c)
                res = backtrack(i+1)
                for c in arr[i]:
                    charSet.remove(c)

            return max(res, backtrack(i+1))

        return backtrack(0)
